File: 01365. How Many Numbers Are Smaller Than the Current Number/solution.cs
/**
 * Problem: 1365. How Many Numbers Are Smaller Than the Current Number
 *
 * Difficulty: Easy
 *
 * Language: C#
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

public class Solution
{
  public int[] SmallerNumbersThanCurrent(int[] nums)
  {
    // Length of the input array
    int arrayLength = nums.Length;

    // Frequency array for numbers 0-100
    int[] frequency = new int[101];
    // Result array to store counts of smaller numbers
    int[] result = new int[arrayLength];

    // Count occurrences of each number in the input array
    for (int i = 0; i < arrayLength; i++)
      frequency[nums[i]]++;

    // Variable to accumulate count of numbers smaller than current value
    int numbersSmallerCount = 0;

    // Convert frequency array to prefix sum of smaller numbers
    for (int value = 0; value < 101; value++)
    {
      // Store current frequency before overwriting
      int currentFrequency = frequency[value];

      // Update frequency[value] to hold count of numbers smaller than value
      frequency[value] = numbersSmallerCount;
      // Add current frequency to accumulator for next iteration
      numbersSmallerCount += currentFrequency;
    }

    // For each number in original array, get count of smaller numbers
    for (int i = 0; i < arrayLength; i++)
      result[i] = frequency[nums[i]];

    // Return the result array
    return result;
  }
}

File: 01365. How Many Numbers Are Smaller Than the Current Number/solution.go
/**
 * Problem: 1365. How Many Numbers Are Smaller Than the Current Number
 *
 * Difficulty: Easy
 *
 * Language: Golang
 *
 * Performance: Runtime - 0 ms (Beats 100%)
 */

func smallerNumbersThanCurrent(nums []int) []int {
	// Frequency array for numbers 0-100
	frequency := make([]int, 101)

	// Result slice to store counts for each position
	result := make([]int, len(nums))

	// Count occurrences of each number
	for i := 0; i < len(nums); i++ {
		frequency[nums[i]]++
	}
	// Build prefix sum to get count of numbers <= current value
	for value := 1; value < 101; value++ {
		frequency[value] += frequency[value-1]
	}

	// For each element, count numbers strictly smaller than it
	for i := 0; i < len(nums); i++ {
		// Get current value at index i
		currentValue := nums[i]

		// If value is 0, no numbers are smaller
		if currentValue == 0 {
			result[i] = 0
		} else {
			// Otherwise, get count of numbers <= (currentValue - 1)
			result[i] = frequency[currentValue-1]
		}
	}

	// Return the result slice with counts of smaller numbers
	return result
}
